pkg/storage/unified/testing/benchmark.go

```diff
@@ -0,0 +1,172 @@
+	NumGroups        int // number of different groups
+	NumResourceTypes int // number of different resource types
+}
+
+// DefaultBenchmarkOptions returns the default benchmark options
+func DefaultBenchmarkOptions() *BenchmarkOptions {
+	return &BenchmarkOptions{
+		NumResources:     1000,
+		Concurrency:      20,
+		NumNamespaces:    1,
+		NumGroups:        1,
+		NumResourceTypes: 1,
+	}
+}
+
+// BenchmarkResult contains the benchmark metrics
+type BenchmarkResult struct {
+	TotalDuration time.Duration
+	WriteCount    int
+	Throughput    float64 // writes per second
+	P50Latency    time.Duration
+	P90Latency    time.Duration
+	P99Latency    time.Duration
+}
+
+// runStorageBackendBenchmark runs a write throughput benchmark
+func runStorageBackendBenchmark(ctx context.Context, backend resource.StorageBackend, opts *BenchmarkOptions) (*BenchmarkResult, error) {
+	if opts == nil {
+		opts = DefaultBenchmarkOptions()
+	}
+
+	// Create channels for workers
+	jobs := make(chan int, opts.NumResources)
+	results := make(chan time.Duration, opts.NumResources)
+	errors := make(chan error, opts.NumResources)
+
+	// Fill the jobs channel
+	for i := 0; i < opts.NumResources; i++ {
+		jobs <- i
+	}
+	close(jobs)
+
+	// Start workers
+	var wg sync.WaitGroup
+
+	// Initialize each group and resource type combination in the init namespace
+	namespace := "ns-init"
+	for g := 0; g < opts.NumGroups; g++ {
+		group := fmt.Sprintf("group-%d", g)
+		for r := 0; r < opts.NumResourceTypes; r++ {
+			resourceType := fmt.Sprintf("resource-%d", r)
+			_, err := writeEvent(ctx, backend, "init", resource.WatchEvent_ADDED,
+				WithNamespace(namespace),
+				WithGroup(group),
+				WithResource(resourceType),
+				WithValue([]byte("init")))
+			if err != nil {
+				return nil, fmt.Errorf("failed to initialize backend: %w", err)
+			}
+		}
+	}
+	// Start workers
+	startTime := time.Now()
+	for workerID := 0; workerID < opts.Concurrency; workerID++ {
+		wg.Add(1)
+		go func(workerID int) {
+			defer wg.Done()
+			for jobID := range jobs {
+				// Calculate a unique ID for this job that's guaranteed to be unique across all workers
+				uniqueID := jobID
+
+				// Generate deterministic and unique resource details
+				namespace := fmt.Sprintf("ns-%d", uniqueID%opts.NumNamespaces)
+				group := fmt.Sprintf("group-%d", uniqueID%opts.NumGroups)
+				resourceType := fmt.Sprintf("resource-%d", uniqueID%opts.NumResourceTypes)
+				// Ensure name is unique by using the global uniqueID
+				name := fmt.Sprintf("item-%d", uniqueID)
+
+				writeStart := time.Now()
+				_, err := writeEvent(ctx, backend, name, resource.WatchEvent_ADDED,
+					WithNamespace(namespace),
+					WithGroup(group),
+					WithResource(resourceType),
+					WithValue([]byte(strings.Repeat("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789", 20)))) // ~1kb
+
+				if err != nil {
+					errors <- err
+					return
+				}
+
+				results <- time.Since(writeStart)
+			}
+		}(workerID)
```

Comment:
nit: passing `workerID` as a function parameter to goroutine is unnecessary since Go 1.22. (Each for iteration has its own `workerID` local variable)